Automattic/woocommerce-payments
# Fetch, cache, and localize for JS onboarding wizard fields data from PHP, collect and prefill MCC

Fixes #6104
Fixes #4853 

#### Changes proposed in this Pull Request

This PR ended up handling two issues in the interest of speed of development and reduced back-and-forth.

First, we use the REST endpoint introduced by [this paired server PR](https://github.com/Automattic/woocommerce-payments-server/pull/3310) to fetch all the fields data in one request (business types and industry/MCC fields, right now). This data is cached in the DB for a week. 

Furthermore, the data is provided directly from PHP for JS consumption rather than being fetched async through a client REST endpoint. By doing it this way, we reduce the client logic complexity (no need for loading states for controls) and increase the client-side UI speed since the data is available immediately. We do end up with a higher HTML payload, but when I considered the current state of the markup (huge, mainly due to the WP media dependency), I decided that the downside in performance is marginal.

Second, we implement the industry/MCC onboarding wizard control with some adjustments to the `GroupedSelectField` component (introduced in [this PR](https://github.com/Automattic/woocommerce-payments/pull/5895)). The adjustments are needed so we are able to use bespoke select options that are custom-ordered (groups and items in a group). The MCC field has these requirements to allow for a focus on the merchant experience rather than programmatic generation.

Here's what the new onboarding prototype MCC control looks like:
![Screenshot 2023-05-02 at 12 01 17](https://user-images.githubusercontent.com/8830539/235624871-6d964cd4-6ae0-4204-a087-f0a1b3db2ff5.png)
![Screenshot 2023-05-02 at 12 01 56](https://user-images.githubusercontent.com/8830539/235624948-4d94c36d-a8a1-4b3b-ad58-a100f35ef4c8.png)

The collected MCC is passed along with the rest of the onboarding data to the server for processing, storage, and prefilling in the Stripe account (implemented in [the paired PR](https://github.com/Automattic/woocommerce-payments-server/pull/3310)).

#### Testing instructions

* On a dev site with WooCommerce Payments, checkout this PR's branch: `add/6104-preload-and-cache-onboarding-wizard-fields-data`
* On the WCpay server used by this client, checkout [the paired PR's](https://github.com/Automattic/woocommerce-payments-server/pull/3310) branch: `add/c6104-single-global-route-for-onboarding-wizard-fields-data`
* Make sure you force the WCPay client to act as disconnected from the server via the [WCPay Dev Tools admin page](http://localhost:8082/wp-admin/admin.php?page=wcpaydev)
* Go to [the onboarding prototype page](http://localhost:8082/wp-admin/admin.php?page=wc-admin&path=%2Fpayments%2Fonboarding-prototype)
* On the third step, you should see the industry control with all the grouped options (labeled "What type of goods or services does your business sell?")
* Select, search, clear the search, close and reopen, select again; basically, smoke test the control
* At the same time, smoke test the business type and structure control (labeled "Where is your business legally registered?")
* Go through the entire onboarding, and you should end up with a Stripe account that has the MCC prefilled (check the Stripe account's dashboard for this); for example, if you chose the "Consulting" option, the Stripe account should have "Professional services (other)" as the industry (see screenshot below)
![Screenshot 2023-05-02 at 12 10 18](https://user-images.githubusercontent.com/8830539/235626843-5503de49-fcc1-410a-ab1a-7564c5eb0e02.png)

-------------------


**Post merge**
